File: examples3d/b3d_trees.rs
//! Ports of box3d's `trees100`/`trees50`/`trees25` benchmarks
//! (`CreateTrees(1|2|4)`, `box3d/shared/benchmarks.c`): 50 "tree" bodies, each
//! a stack of 22 convex cylinders, spun onto a sinusoidal wave-mesh ground.
//! The scale selects the ground tessellation density.

use rapier_testbed3d::TestbedViewer;
use rapier3d::prelude::*;
use std::f32::consts::PI;

pub async fn run100(viewer: &mut TestbedViewer) -> anyhow::Result<()> {
    run(viewer, 1).await
}

pub async fn run50(viewer: &mut TestbedViewer) -> anyhow::Result<()> {
    run(viewer, 2).await
}

pub async fn run25(viewer: &mut TestbedViewer) -> anyhow::Result<()> {
    run(viewer, 4).await
}

async fn run(viewer: &mut TestbedViewer, scale: usize) -> anyhow::Result<()> {
    // box3d's `b3DefaultWorldDef`: gravity (0, -10, 0). box3d steps at dt = 1/60
    // with 4 solver substeps, matching rapier's defaults.
    let mut world = PhysicsWorld::new();
    world.gravity = Vector::new(0.0, -10.0, 0.0);
    create_trees(&mut world, scale);

    viewer.set_world(&mut world);
    viewer.look_at(Vec3::new(0.0, 30.0, 140.0), Vec3::new(0.0, 15.0, 0.0));

    while viewer.render_frame(&mut world).await {
        if viewer.simulating() {
            world.step();
        }
    }
    Ok(())
}

/// Port of box3d's `CreateTrees` (`box3d/shared/benchmarks.c`). `scale` is
/// 1/2/4 for the trees100/50/25 variants. Builds a sinusoidal wave-mesh ground
/// and 50 "tree" bodies, each a stack of 22 convex cylinders, spun about Z with
/// alternating direction (release settings, `tilt = 0`).
fn create_trees(world: &mut PhysicsWorld, scale: usize) {
    let x_count = scale * 150;
    let z_count = scale * 200;
    let cell_width = 1.0 / scale as f32;
    let amplitude = 0.4;
    let row_hz = 0.05;
    let column_hz = 0.1;

    let (vertices, indices) =
        create_wave_mesh(x_count, z_count, cell_width, amplitude, row_hz, column_hz);
    let ground = world.insert_body(RigidBodyBuilder::fixed());
    world.insert_collider(
        ColliderBuilder::trimesh(vertices, indices).unwrap(),
        Some(ground),
    );

    // The 22 stacked cylinders shared by every tree body. Building each hull
    // once (as a shared shape) lets all 50 trees reuse the geometry and lets
    // the renderer instance the repeated hulls.
    let hull_count = 22usize;
    let mut hulls = Vec::with_capacity(hull_count);
    let mut y = 1.0f32;
    let mut r = 0.75f32;
    let l = 1.5f32;
    for _ in 0..hull_count {
        let points = create_cylinder(l + 2.0 * r, r, y - r, 6);
        hulls.push(SharedShape::convex_hull(&points).unwrap());
        y += l + 2.0 * r;
        r *= 0.95;
    }

    let body_count = 50i32;
    let mut angular_velocity = -0.5f32;
    let mut z = -70.0f32;
    for body_index in 0..body_count {
        let position = Vector::new(0.0, 1.0, z);
        // box3d applies gyroscopic torque to every body every substep (see
        // `b3IntegrateVelocitiesTask`, "improves the simulation of long skinny
        // bodies"). rapier gates that term behind this flag, off by default, so
        // enable it here or the ~1300:1-inertia trees tumble along wrong paths.
        let handle = world.insert_body(RigidBodyBuilder::dynamic().translation(position));
        for hull in &hulls {
            world.insert_collider(
                ColliderBuilder::new(hull.clone())
                    .density(1.0)
                    .friction(0.9),
                Some(handle),
            );
        }

        let velocity_scale = 0.5 + (0.5 * body_index as f32) / body_count as f32;
        let center = world.bodies[handle].center_of_mass();
        let omega = Vector::new(0.0, 0.0, velocity_scale * angular_velocity);
        let v = omega.cross(center - position);
        let body = &mut world.bodies[handle];
        body.set_angvel(omega, true);
        body.set_linvel(v, true);

        z += 3.0;
        angular_velocity = -angular_velocity;
    }
}

/// box3d `b3CreateWaveMesh` (`src/mesh.c`): a grid whose height is a product of
/// two sinusoids, used as the "trees" benchmark ground.
fn create_wave_mesh(
    x_count: usize,
    z_count: usize,
    cell_width: f32,
    amplitude: f32,
    row_frequency: f32,
    column_frequency: f32,
) -> (Vec<Vector>, Vec<[u32; 3]>) {
    let omega_z = 2.0 * PI * row_frequency * cell_width;
    let omega_x = 2.0 * PI * column_frequency * cell_width;
    grid_mesh(x_count, z_count, cell_width, |ix, iz| {
        amplitude * (omega_x * ix as f32).sin() * (omega_z * iz as f32).sin()
    })
}

/// Shared vertex/index grid builder for the box3d grid/wave meshes. `height`
/// yields the Y coordinate at grid cell `(ix, iz)`.
fn grid_mesh(
    x_count: usize,
    z_count: usize,
    cell_width: f32,
    height: impl Fn(usize, usize) -> f32,
) -> (Vec<Vector>, Vec<[u32; 3]>) {
    let x_width = cell_width * x_count as f32;
    let z_width = cell_width * z_count as f32;

    let mut vertices = Vec::with_capacity((x_count + 1) * (z_count + 1));
    let mut x = -0.5 * x_width;
    for ix in 0..=x_count {
        let mut z = -0.5 * z_width;
        for iz in 0..=z_count {
            vertices.push(Vector::new(x, height(ix, iz), z));
            z += cell_width;
        }
        x += cell_width;
    }

    let mut indices = Vec::with_capacity(2 * x_count * z_count);
    for ix in 0..x_count {
        for iz in 0..z_count {
            let i1 = (iz + (z_count + 1) * ix) as u32;
            let i2 = i1 + 1;
            let i3 = i2 + (z_count as u32 + 1);
            let i4 = i3 - 1;
            indices.push([i1, i2, i3]);
            indices.push([i3, i4, i1]);
        }
    }
    (vertices, indices)
}

/// box3d `b3CreateGridMesh` (`src/mesh.c`): a flat `x_count * z_count` grid.
fn create_grid_mesh(
    x_count: usize,
    z_count: usize,
    cell_width: f32,
) -> (Vec<Vector>, Vec<[u32; 3]>) {
    grid_mesh(x_count, z_count, cell_width, |_, _| 0.0)
}

/// box3d `b3CreateCylinder` (`src/hull.c`): `2 * sides` points forming a
/// cylinder of the given `height`/`radius`, its base at `y_offset`, aligned
/// with the Y axis. Returned as a convex point cloud (rapier builds the hull).
fn create_cylinder(height: f32, radius: f32, y_offset: f32, sides: usize) -> Vec<Vector> {
    let mut points = Vec::with_capacity(2 * sides);
    let delta_alpha = 2.0 * PI / sides as f32;
    let mut alpha = 0.0f32;
    for _ in 0..sides {
        let (sin_a, cos_a) = alpha.sin_cos();
        points.push(Vector::new(radius * cos_a, y_offset, radius * sin_a));
        points.push(Vector::new(
            radius * cos_a,
            y_offset + height,
            radius * sin_a,
        ));
        alpha += delta_alpha;
    }
    points
}

File: c/testbed/examples3d/b3d_trees.c
/* Port of examples3d/b3d_trees.rs. */
#include "testbed.h"
#include "rapier_helpers.h"
#include "rapier_math.h"

static void createCylinder(R3Real height, R3Real radius, R3Real yOffset, size_t sides,
                           R3Vector *points) {
    const R3Real deltaAlpha = 2 * R3_PI / sides;
    R3Real alpha = 0;
    for (size_t i = 0; i < sides; ++i) {
        const R3Real sinA = sin(alpha), cosA = cos(alpha);
        points[2 * i] = r3Vector(radius * cosA, yOffset, radius * sinA);
        points[2 * i + 1] = r3Vector(radius * cosA, yOffset + height, radius * sinA);
        alpha += deltaAlpha;
    }
}

static void createWaveMesh(size_t xCount, size_t zCount, R3Real cellWidth, R3Real amplitude,
                           R3Real rowFrequency, R3Real columnFrequency, R3Vector *vertices,
                           uint32_t *indices) {
    const R3Real omegaZ = 2 * R3_PI * rowFrequency * cellWidth;
    const R3Real omegaX = 2 * R3_PI * columnFrequency * cellWidth;
    R3Real x = -.5 * cellWidth * xCount;
    for (size_t ix = 0; ix <= xCount; ++ix) {
        R3Real z = -.5 * cellWidth * zCount;
        for (size_t iz = 0; iz <= zCount; ++iz) {
            vertices[ix * (zCount + 1) + iz] =
                r3Vector(x, amplitude * sin(omegaX * ix) * sin(omegaZ * iz), z);
            z += cellWidth;
        }
        x += cellWidth;
    }
    size_t next = 0;
    for (size_t ix = 0; ix < xCount; ++ix) {
        for (size_t iz = 0; iz < zCount; ++iz) {
            const uint32_t i1 = iz + (zCount + 1) * ix, i2 = i1 + 1;
            const uint32_t i3 = i2 + zCount + 1, i4 = i3 - 1;
            indices[next++] = i1;
            indices[next++] = i2;
            indices[next++] = i3;
            indices[next++] = i3;
            indices[next++] = i4;
            indices[next++] = i1;
        }
    }
}

void b3dTreesRun(Testbed *testbed, size_t scale) {
    R3World *world = r3NewWorld();
    r3SetGravity(world, r3Vector(0, -10, 0));
    const size_t xCount = scale * 150, zCount = scale * 200;
    const size_t vertexCount = (xCount + 1) * (zCount + 1), triangleCount = 2 * xCount * zCount;
    R3Vector *vertices = malloc(vertexCount * sizeof(*vertices));
    uint32_t *indices = malloc(triangleCount * 3 * sizeof(*indices));
    if (!vertices || !indices) {
        abort();
    }
    createWaveMesh(xCount, zCount, 1.0 / scale, .4, .05, .1, vertices, indices);
    {
        R3RigidBodyDesc rigidBody = r3FixedRigidBodyDesc();
        rigidBody.position.translation = r3Vector(0, 0, 0);
        rigidBody.canSleep = !testbed->noSleep;
        R3ColliderDesc collider = r3DefaultColliderDesc();
        r3ShapeDesc_SetTrimesh(&collider.shape, (R3VectorView){vertices, vertexCount},
                              (R3TriangleView){(const R3Triangle *)indices, triangleCount}, 0);
        R3RigidBodyHandle rigidBodyHandle = r3InsertRigidBody(world, &rigidBody);
        r3InsertCollider(rigidBodyHandle, &collider);
    }
    free(vertices);
    free(indices);
    R3SharedShape *hulls[22] = {0};
    R3Real y = 1, r = .75;
    const R3Real l = 1.5;
    for (size_t i = 0; i < TB_COUNT(hulls); ++i) {
        R3Vector points[12];
        createCylinder(l + 2 * r, r, y - r, 6, points);
        hulls[i] = r3ConvexHullSharedShape((R3VectorView){points, TB_COUNT(points)});
        y += l + 2 * r;
        r *= .95;
    }

    R3Real angularVelocity = -.5, z = -70;
    const int bodyCount = 50;
    for (int bodyIndex = 0; bodyIndex < bodyCount; ++bodyIndex) {
        const R3Vector position = r3Vector(0, 1, z);
        R3RigidBodyDesc builder = r3DynamicRigidBodyDesc();
        builder.position.translation = position;
        builder.canSleep = !testbed->noSleep;
        R3RigidBodyHandle handle = r3InsertRigidBody(world, &builder);

        for (size_t i = 0; i < TB_COUNT(hulls); ++i) {
            R3ColliderDesc collider = r3DefaultColliderDesc();
            collider.shape.kind = R3_SHAPE_DESC_SHARED;
            collider.shape.sharedShape = hulls[i];
            collider.density = 1;
            collider.friction = .9;
            r3InsertCollider(handle, &collider);
        }
        const R3Real velocityScale = .5 + .5 * bodyIndex / bodyCount;

        R3Vector center = r3RigidBody_CenterOfMass(handle);
        const R3Vector omega = r3Vector(0, 0, velocityScale * angularVelocity);
        const R3Vector velocity = r3VectorCross(omega, r3VectorSub(center, position));
        r3RigidBody_SetAngvel(handle, omega, 1);
        r3RigidBody_SetLinvel(handle, velocity, 1);
        z += 3;
        angularVelocity = -angularVelocity;
    }
    for (size_t i = 0; i < TB_COUNT(hulls); ++i) {
        r3FreeSharedShape(hulls[i]);
    }
    tbCamera(testbed, 0, 30, 140, 0, 15, 0);
    tbSetWorld(testbed, world);

    while (tbRenderFrame(testbed, &world)) {
        if (tbSimulating(testbed)) {
            r3Step(world, NULL, NULL);
        }
    }
    r3FreeWorld(world);
}
